New protocol documentation. (#380)
LiskHQ/lisk-docs: modules/ROOT/pages/0-introduction.adoc

= Introduction

> _This documentation only describes the general features of the Lisk protocol version *XXX*. Notice that this ia a future version of the Lisk protocol. For a detailed technical explanation go [here]._

Reviewer: ```suggestion
_This documentation only describes the general features of the Lisk protocol version *XXX*. Notice that this is a future version of the Lisk protocol. For a detailed technical explanation go [here]._
```

> _This documentation only describes the general features of the Lisk protocol version *XXX*. Notice that this is a future version of the Lisk protocol. For a detailed technical explanation go [here]._

Author: ```suggestion
_This documentation gives an overview of the Lisk protocol, the set of rules followed by a blockchain created with the Lisk SDK using the default configuration. A blockchain that uses custom constants or adds custom transactions (see the [Lisk SDK documentation](https://lisk.io/documentation/lisk-sdk/index.html) for details) may therefore deviate from the protocol described here. The differences between the Lisk Mainnet protocol and Lisk protocol is covered in the last section of the documentation. Note that the documentation describes the Lisk protocol after the completion of the ["Network Longevity" roadmap objective](https://lisk.io/roadmap), which is currently not yet implemented. 

The aim of this documentation is to provide a high-level overview of the protocol. A complete technical specification of most parts of the protocol is given in the respective [Lisk Improvemet Proposals](https://github.com/LiskHQ/lips)._
```

Author: The documentation is supposed to be published soon. Therefore, we shouldn't use placeholders for things that are not defined yet or not written yet.

== What is a Blockchain?

At its core, a [#index-blockchain-1]#*blockchain*# is a database shared across multiple computers.
Operations on the database are called *transactions*.
The data contained in a transaction specifies the operation to be performed.

> In an [#index-account_based-1]#*account-based*# blockchain (such as Ethereum, see https://en.wikipedia.org/wiki/Ethereum[Wikipedia page]), the validity of a new transaction is checked against a database of user accounts. An account is a portion of the database identified by the account [#index-address-1]#*address*#.

Author: I would remove the example in brackets here.

Reviewer: Only the ethereum one? What is the reason?

This is opposed to [#index-UTXO_based-1]#*UTXO-based*# blockchains (such as Bitcoin, see https://en.wikipedia.org/wiki/Bitcoin[Wikipedia page]), where there is no concept of account and the validity of a transaction is based on referenced past transactions.
The cost of processing the transaction and updating the blockchain is specified in the transaction fee and paid using [#index-tokens-1]#*tokens*#.
Transactions are shared among users connected in a [#index-P2P_network-1]#*P2P network*# and are grouped together into [#index-blocks-1]#*blocks*#.
The validity of a block (and all transactions within it) is determined against the accounts state.

> The [#index-consensus_protocol-1]#*consensus protocol*# specifies who has the right to [#index-forge-1]#*forge*# a new block and how agreement on the validity of a block is reached.

Author: ```suggestion
The [#index-consensus_protocol-1]#*consensus protocol*# specifies who has the right to [#index-forge-1]#*forge*# a new block and how agreement on blocks is reached.
```

Author: The validity rules for blocks should be all the same and consensus is rather about reaching agreement on which block belongs to the blockchain rather than whether a block is valid.

Reviewer: done

After the block has been forged, it is added to the blockchain, and all state changes induced by the block transactions are applied.

Starting from the initial [#index-genesis_block-1]#*genesis block*# (the initial state of the database), the current [#index-state_of_the_blockchain-1]#*state of the blockchain*# can be calculated by successively applying all transactions contained in the blocks.
In this sense, _defining the blockchain to be the history of the underlying database or the ordered collection of blocks, is equivalent_.

> insert big diagram from https://projects.invisionapp.com/share/SRVCPRZ7F83#/screens/398517617_Infographic_-_Contracted[here]

Author: I think it would be better to put the diagram at the end of the "Overview of the Lisk Protocol" section. We could also add an introductory sentence such as "The main characteristics of the Lisk protocol are shown in the following diagram.".

Reviewer: done

== Overview of the Lisk Protocol

The Lisk protocol is the formal set of rules that defines a blockchain created with the Lisk SDK.
The Lisk protocol defines an account-based blockchain with a fixed set of [#index-transaction_types-1]#*default transaction types*# that are used to modify the accounts state.
The fees for those transactions are paid with the chain native token.

> Transactions are gathered in blocks by [#index-delegates-1]#*delegates*#, which are selected using a [#index-delegated_proof_of_stake-1]#*delegated proof-of-stake*# mechanism.

Author: ```suggestion
Transactions are gathered in blocks by [#index-delegates-1]#*delegates*#, which are selected using a [#index-delegated_proof_of_stake-1]#*Delegated Proof-of-Stake (DPoS)*# mechanism.
```

Author: The term in brackets is optional.

Author: Same in the accounts section.

Reviewer: done

> The Lisk Protocol is the formal set of rules that defines the Lisk blockchain. Lisk is an [#index-account_based-1]#*account-based*# blockchain, where the validity of a new transaction is checked against a database of user accounts. This is opposed to [#index-UTXO_based-1]#*UTXO-based*# blockchains (like Bitcoin [https://en.wikipedia.org/wiki/Bitcoin[1]]), where there is no concept of account and the validity of a transaction is based on referenced past transactions. The Lisk protocol defines a fixed set of [#index-transaction_types-1]#*transaction types*# that are used to modify the account state. The fees for those transactions are paid with the native token *LSK*. Each LSK is further subdivided into 10^8^ [#index-beddows-1]#*Beddows*#. Transactions are gathered in blocks by [#index-delegates-1]#*delegates*#, which are selected using a [#index-delegated_proof_of_stake-1]#*delegated proof-of-stake*# mechanism. Delegates agree on blocks using the [#index-lisk_bft-1]#*Lisk-BFT*# consensus algorithm.

Reviewer: done

Delegates agree on blocks using the [#index-lisk_bft-1]#*Lisk-BFT*# consensus algorithm.

This documentation is structured as follows:

. link:1-accounts.adoc[*Accounts*]: In this section, user accounts forming the underlying database are covered.
. link:2-transactions.adoc[*Transactions*]: This section outlines the default transactions present in the Lisk protocol.
. link:3-blocks.adoc[*Blocks*]: This section describes the actual data stored in a block and how new blocks are created.
. link:4-consensus-algorithm.adoc[*Consensus Algorithm*]: This section illustrates how the delegates eligible to create blocks are selected and how they reach agreement on new blocks.

> . link:5-network.adoc[*Network*]: This section describes the Lisk P2P network used to exchange information such as blocks or transactions between nodes.

Author: Section 6 and 7 are missing. I would also say that "Lisk Mainnet" should come before the appendix.

Reviewer: done

> = Introduction

Reviewer 2: We need to state that this is the future protocol and not the current one.

Reviewer: I thought the protocol version would be implied by the docs version, and we would keep past version available. Maybe I'm wrong. I'm adding a placeholder (I don't know which protocol version we are documenting here)

Reviewer 2: I'm also not sure which version number it will be. I guess 5.0. But I still think that it's worth to add that this a FUTURE protocol. I think that many readers do not know which protocol version is currently implemented in the SDK or used on the mainchain.

Reviewer: Maxime suggested to otherwise mention the roadmap milestones, so that people can check in that page if they have been implemented already.

> 

Reviewer 2: We need to state that this is the future protocol and not the current one.

insert big diagram from https://projects.invisionapp.com/share/SRVCPRZ7F83#/screens/398517617_Infographic_-_Contracted[here]

== Overview of the Lisk Protocol

The Lisk Protocol is the formal set of rules that defines a blockchain created with the Lisk SDK.
The Lisk protocol defines an account-based blockchain with a fixed set of [#index-transaction_types-1]#*default transaction types*# that are used to modify the accounts state.
The fees for those transactions are paid with the chain native token.
Transactions are gathered in blocks by [#index-delegates-1]#*delegates*#, which are selected using a [#index-delegated_proof_of_stake-1]#*delegated proof-of-stake*# mechanism.
Delegates agree on blocks using the [#index-lisk_bft-1]#*Lisk-BFT*# consensus algorithm.

This documentation is structured as follows:

. link:1-accounts.adoc[*Accounts*]: In this section, user accounts forming the underlying database are covered.
. link:2-transactions.adoc[*Transactions*]: This section outlines the default transactions present in the Lisk protocol.
. link:3-blocks.adoc[*Blocks*]: This section describes the actual data stored in a block and how new blocks are created.

> . link:4-consensus-algorithm.adoc[*Consensus Algorithm*]: In this section illustrates how the delegates eligible to create blocks are selected and how they reach agreement on new blocks.

Reviewer 2: ```suggestion
. link:4-consensus-algorithm.adoc[*Consensus Algorithm*]: This section illustrates how the delegates eligible to create blocks are selected and how they reach agreement on new blocks.
```

Reviewer 3: Agreed.

Reviewer: fixed
